core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -84,19 +76,65 @@ public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpira
     deleteFiles(manifestListsToDelete, "manifest list");
   }
 
-  private Set<ManifestFile> readManifests(Set<Snapshot> snapshots) {
-    Set<ManifestFile> manifestFiles = Sets.newHashSet();
-    for (Snapshot snapshot : snapshots) {
-      try (CloseableIterable<ManifestFile> manifestFilesForSnapshot = readManifestFiles(snapshot)) {
-        for (ManifestFile manifestFile : manifestFilesForSnapshot) {
-          manifestFiles.add(manifestFile.copy());
-        }
-      } catch (IOException e) {
-        throw new RuntimeIOException(
-            e, "Failed to close manifest list: %s", snapshot.manifestListLocation());
-      }
+  private Set<ManifestFile> currentManifests(
+      Set<Snapshot> snapshots, Set<ManifestFile> manifestsToDelete) {
+    Set<ManifestFile> currentManifests = ConcurrentHashMap.newKeySet();
+    if (manifestsToDelete.isEmpty()) {
+      return currentManifests;
```

Comment:
Correctness Question / Naming Nit -This seems somewhat odd to me. The way the function is used, this might produce correct results.

But looking at the function name / signature directly and the first few lines of code, it's strange to return an empty set as the "current manifests" if the manifests to delete are empty.

Generally speaking, that's not true to say that if there are no manifests to delete, there are no current manifests.

Maybe the function needs to be renamed or the local variable name needs to be renamed?

Again, the usage is very possibly correct, but the name of the function and logic in the base case of empty manifests seems off.

Is it supposed to return an empty set when `snapshots` is empty?

Reply:
I think this is ultimately a naming problem, I struggled to find a concise name since it's really doing multiple things:

1.) The function is reading the current manifests and pruning out the candidate manifests to delete at the same time. The reason it is done this way is ultimately for this procedure we are optimizing for memory usage, and the current manifests are expected to be a larger set on average compared to the manifests to delete. 

2.) Following 1. what this means is the procedure should aim to minimize the amount of the current table state to read. The procedure can stop reading manifests in the current table state if all the candidate manifests have been pruned.

>Generally speaking, that's not true to say that if there are no manifests to delete, there are no current manifests.

3.) You're definitely correct here but from a "deletion correctness"  standpoint this state is still good because if there are no manifests to delete this means that there are no data files we can actually delete and subsequent findFilesToDelete operations are guaranteed to be no-ops.

Reply:
I renamed to pruneManifestsToDelete and added some inliine comments; although still not great naming since the return set is the current manifests. 

I think another way to make it more explicit and avoid the mutating of state of the candidate manifests to delete is to have the function return a Pair<Set<ManifestFile>, Set<ManifestFIle>> , where the first entry is the current manifests and the second is the manifests that can safely be removed, and these 2 sets can be used when determining the data files to delete later on. I initially avoided this since the Pair seemed a bit more complex than just mutating the state, 

Let me know which approach you find preferable @kbendick @rdblue

Reply:
Ha, I just commented above that I'd prefer the Pair option. This is just an awkward case.

What about adding a callback to do this? Like this:

```java
private Set<ManifestFile> pruneReferencedManifests(Set<ManifestFile> deleteCandidates, Set<Snapshot> currentSnapshots, Consumer<ManifestFile> currentManifestCallback) {
  Set<ManifestFile> candidateSet = ConcurrentHashMap.keySet();
  candidateSet.addAll(deleteCandidates); // ensure the incoming set is thread safe!
  for (ManifestFile manifest : manifestFiles) {
    candidateSet.remove(manifest);
    currentManifestCallback.accept(manifest);
  }
  ...
  return candidateSet;
}

// when calling pruneReferencedManifests:
Set<ManifestFile> currentManifests = ConcurrentHashMap.keySet();
Set<ManifestFile> manifestsToDelete = pruneReferencedManifests(deleteCandidates, currentSnapshots, currentManifests::add);
```

Reply:
Updated to use the callback approach.